### PR TITLE
Migration to Harmony / RedirectionFramework

### LINKED ISSUE #119: Migration to Harmony
_As we know there are serious problems with overriding our methods by other mods._

__@VictorPhilipp__ did great job with implementation in separate branch. It is more than __5 patches__ behind of our current version (about __140 commits__ :smile: ) but it has some other changes to citizen manager, Flags and other things that looks promissing but not released yet. 
I have to test if everything it in place and _'working'_.

I will try to merge our latest changes and push it to new branch, probably 1.11.0-pre-alpha to give you opportunity to do more tests on different environments, check compatibility with savegames(loading/saving) etc. 

Maybe I'll even add compiled dlls to speed up testing.

### PR DESCRIPTION
**Pull request by VictorPhilipp**
Fixes #119

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,6 +1,9 @@
...
 	path = TLM/OptionsFramework
-	url = ../OptionsFramework.git
+	url = https://github.com/VictorPhilipp/OptionsFramework.git
 [submodule "TLM/CSUtil.CameraControl"]
 	path = TLM/CSUtil.CameraControl
-	url = ../CSUtil.CameraControl.git
+	url = https://github.com/VictorPhilipp/CSUtil.CameraControl.git
+[submodule "TLM/TLM/RedirectionFramework"]
+	path = TLM/TLM/RedirectionFramework
+	url = https://github.com/VictorPhilipp/RedirectionFramework.git
```
Comment (VictorPhilipp) [inline .gitmodules]: @krzychu124 would need to fork this repo

Comment (krzychu124) [reply .gitmodules]: Done 😉

File: TLM/TLM/Manager/Impl/ExtVehicleManager.cs [hunk 2]
```diff
@@ -639,7 +639,7 @@ public class ExtVehicleManager : AbstractCustomManager, IExtVehicleManager {
 				case VehicleInfo.VehicleType.Car:
 					if (ai is PassengerCarAI)
 						return ExtVehicleType.PassengerCar;
-					if (ai is AmbulanceAI || ai is FireTruckAI || ai is PoliceCarAI || ai is HearseAI || ai is GarbageTruckAI || ai is MaintenanceTruckAI || ai is SnowTruckAI || ai is WaterTruckAI || ai is DisasterResponseVehicleAI || ai is ParkMaintenanceVehicleAI) {
+					if (ai is AmbulanceAI || ai is FireTruckAI || ai is PoliceCarAI || ai is HearseAI || ai is GarbageTruckAI || ai is MaintenanceTruckAI || ai is SnowTruckAI || ai is WaterTruckAI || ai is DisasterResponseVehicleAI || ai is ParkMaintenanceVehicleAI || ai is PostVanAI) {
```
Comment (aubergine10) [inline TLM/TLM/Manager/Impl/ExtVehicleManager.cs]: Is there also a CargoTruckAI variant of the post van? (Spawned from Post Sorting Office I think)

Example assets in workshop:

* [Posti Truck](https://steamcommunity.com/sharedfiles/filedetails/?id=1650066257)
* [Volea Truck](https://steamcommunity.com/sharedfiles/filedetails/?id=1559503051)

Comment (krzychu124) [reply TLM/TLM/Manager/Impl/ExtVehicleManager.cs]: I think that different AI type is used when post van is delivering mail and other when transporting mail outside city 😉

Comment (aubergine10) [reply TLM/TLM/Manager/Impl/ExtVehicleManager.cs]: Ah, yes, probably.

With the post truck still be a `VehicleInfo.VehicleType.Car` though?

Comment (krzychu124) [reply TLM/TLM/Manager/Impl/ExtVehicleManager.cs]: I am not sure if I understand your question 😕 
This switch case was made for determining TMPE Extended type using vanilla VehicleType and AI of that vehicle. 
I am not sure but `Post van` should use `PostVanAI` and `Post truck`(which goes outside of the city) uses `CargoTruckAI`. 
Nonetheless it's still vanilla `VehicleType.Car` and I can't see an other better match 😉

**Comment by krzychu124** [comment]
Little issue:
```cs
The following methods were overriden by another mod:
	<Harmony> HumanAI.ArriveAtDestination with 3 parameters (HumanAI, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null)
```
Mod list:
- ModTools
- Network Extensions 2
- Loading Screen Mod
- 81 Tiles (Fixed for C:S 1.2+) - disabled
- TM:PE 11.0-alpha

**Comment by krzychu124** [comment]
A lot of different features tested and so far no issues :+1: , but long way to finish 😄

**Comment by aubergine10** [comment]
I'll be testing this later today with about 120 active mods.

**Comment by aubergine10** [comment]
Getting loads of errors when trying to build locally...
errors

**Comment by aubergine10** [comment]
Down to 13 errors when trying to build, namely it's saying it can't find any of these files:

redirect

Any ideas?

**Comment by krzychu124** [comment]
You have to update your local  submodules

**Comment by aubergine10** [comment]
How do I do that? Normally github desktop automatically adds/updates the modules. For some reason it's not doing that with this PR or this module...

**Comment by krzychu124** [comment]
Use command line, powershell or gitbash if you have :wink:

**Comment by aubergine10** [comment]
Only issue I've found so far is that the lock menu & button options in mod options screen, when accessed from main menu, throw errors:

```
NullReferenceException: Object reference not set to an instance of an object
  at TrafficManager.State.Options.onLockButtonChanged (Boolean newValue) [0x00000] in <filename unknown>:0 
  at UIHelper+<AddCheckbox>c__AnonStorey0.<>m__0 (ColossalFramework.UI.UIComponent c, Boolean isChecked) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UICheckBox.OnCheckChanged () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UICheckBox.set_isChecked (Boolean value) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UICheckBox.OnClick (ColossalFramework.UI.UIMouseEventParameter p) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.OnClick (ColossalFramework.UI.UIMouseEventParameter p) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIInput+MouseHandler.ProcessInput (IInputTranslator translator, Ray ray, ColossalFramework.UI.UIComponent component, Boolean retainFocusSetting) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIInput.ProcessMouseInput () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIInput.Update () [0x00000] in <filename unknown>:0 
```

I'm also getting lots of these in the log that weren't there before:

```
Assembly 'TrafficManager, Version=1.0.7080.36181, Culture=neutral, PublicKeyToken=null' resolved to ''  [Serialization]
```

But other than those two minor things...

```
[Info] @ 3857630232 ThreadingExtension.OnBeforeSimulationFrame: First frame detected. Checking detours.
[Info] @ 3857671980 ThreadingExtension.OnBeforeSimulationFrame: First frame detected. Detours checked. Result: 0 missing detours
```

Didn't spot any issues in-game so far. That's with >100 active mods (asterisk = enabled): [mods.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3260209/mods.txt)

**Comment by VictorPhilipp** [comment]
Thanks for the testing! I fixed the null reference issue.

File: TLM/TLM/Constants.cs [hunk 3]
```diff
@@ -8,6 +8,7 @@
 namespace TrafficManager {
 	public static class Constants {
 		public static readonly bool[] ALL_BOOL = new bool[] { false, true };
+		public const float BYTE_TO_FLOAT_OFFSET_CONVERSION_FACTOR = 0.003921569f;
```
Comment (kvakvs) [inline TLM/TLM/Constants.cs]: Magical constant is terrible. One can easily see that it is 1.0f / 255f

Comment (krzychu124) [reply TLM/TLM/Constants.cs]: Yes, but it is still better than dividing 1f/255f each time

Comment (aubergine10) [reply TLM/TLM/Constants.cs]: It's a static class, so surely it will only do the actual calculation at compile time?

Comment (kvakvs) [reply TLM/TLM/Constants.cs]: Compilers know exactly when a part of expression can be calculated in the compile time, you don't need to pre-calculate for them, it is important for whoever will read it later, to understand the meaning of this constant. I will change it later when i visit this file and find it, don't bother fixing it right now.

Comment (aubergine10) [reply TLM/TLM/Constants.cs]: tbh, even if it's changed to `1f/255f` (which we should do IMO), I still wouldn't have a clue what it does or how it works -- is "byte to float offset conversion factor" something that's commonly understood?

Comment (krzychu124) [reply TLM/TLM/Constants.cs]: it's simple: for each byte (0-255) you will get fraction of 1f
- 0(byte) == 0f, 
- 255(byte) == 1f

Comment (aubergine10) [reply TLM/TLM/Constants.cs]: Yup, that would be helpful, even if comment was just a URL to a stackoverflow (or whatever) explaining what's going on.

**Comment by aubergine10** [comment]
Testing in-game seems OK to me, and seems to be save/loading successfully.

**Pull request merged**